packages/langium/src/workspace/index-manager.ts

```diff
@@ -106,38 +108,53 @@ export class DefaultIndexManager implements IndexManager {
         return stream(result);
     }
 
-    allElements(nodeType = ''): Stream<AstNodeDescription> {
-        if (!this.globalScopeCache.has('')) {
-            this.globalScopeCache.set('', Array.from(this.simpleIndex.values()).flat());
+    allElements(nodeType?: string, uris?: Set<string>): Stream<AstNodeDescription> {
+        const allUris = this.documents.all.map(doc => doc.uri.toString());
+        const arrays: AstNodeDescription[][] = [];
+        for (const uri of allUris) {
+            if (!uris || uris.has(uri)) {
+                arrays.push(this.getFileDescriptions(uri, nodeType));
+            }
         }
+        return stream(arrays).flat();
```

Comment:
I would use the key set of simpleIndex instead of the LangiumDocuments service. And avoid intermediate arrays.
```typescript
const allUris = this.simpleIndex.keys();
return stream(allUris)
    .filter(uri => !uris || uris.has(uri))
    .map(uri => this.getFileDescriptions(uri, nodeType))
    .flat();
```